### PR TITLE
full dpo

### PR DESCRIPTION
**Pull request by jxmsML**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
*

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: recipes/full_dpo_distributed.py [hunk 1]
```diff
@@ -0,0 +1,970 @@
...
+            enable_activation_offloading=self._enable_activation_offloading,
+            custom_sharded_layers=cfg.get("custom_sharded_layers", None),
+            fsdp_cpu_offload=cfg.get("fsdp_cpu_offload", False),
+            reshard_after_forward=cfg.get("fsdp_reshard_after_forward", True),
+            model_state_dict=checkpoint_dict[training.MODEL_KEY],
+            ac_mode=cfg.get("ac_mode", None),
+            ac_option=cfg.get("ac_option", None),
+        )
+        log.info("Loading reference model")
+        self._ref_model = self._setup_model(
```
Comment (SalmanMohammadi) [inline recipes/full_dpo_distributed.py:304]: You might need to make sure dropout is explictly disabled in these models so it doesn't cause issues when comparing logprobs between the reference and policy models.

File: recipes/full_dpo_distributed.py [hunk 2]
```diff
@@ -0,0 +1,970 @@
...
+                policy_rejected_logits_mean = policy_rejected_logits.detach().mean()
+
+                # deleting logits here helps reduce (peak) memory usage - we only need them for metric logging
+                del policy_chosen_logits, policy_rejected_logits
+
+                if isinstance(self._loss_fn, SimPOLoss):
+                    loss, chosen_rewards, rejected_rewards = self._loss_fn(
+                        policy_chosen_log_probs, policy_rejected_log_probs
+                    )
+                else:
```
Comment (SalmanMohammadi) [inline recipes/full_dpo_distributed.py:845]: We're going to be deprecating non-DPO losses in our DPO recipes so we can take it out and simplify this whole section.

File: recipes/full_dpo_distributed.py [hunk 3]
```diff
@@ -0,0 +1,970 @@
...
+            concatenated_labels[:len_chosen],
+            # see :class:`~torchtune.modules.rlhf.loss.dpo.SimPOLoss`
+            return_average_logprobs=isinstance(self._loss_fn, SimPOLoss),
+        )
+
+        rejected_log_probs = rlhf.get_batch_log_probs(
+            all_logits[len_chosen:],
+            concatenated_labels[len_chosen:],
+            # see :class:`~torchtune.modules.rlhf.loss.dpo.SimPOLoss`
+            return_average_logprobs=isinstance(self._loss_fn, SimPOLoss),
```
Comment (SalmanMohammadi) [inline recipes/full_dpo_distributed.py:783]: Similar to my comment below about SimPO

**Comment by SalmanMohammadi** [comment]
Hey @jxmsML! Thanks so much for opening this. Overall it looks great - just a couple comments for things to look out for. I know this is still in draft so happy to revisit when you're ready for review - it'd be great to see some training outputs and an example config or two (I see you've added one in the recipe registry) at some point.

**Comment by sam-pi** [comment]
Hello, this is great! Thank you, this helped me to get full DPO working, and I ended up just completing the remaining actions in the PR with training examples, configs, and tests. Would it make sense for me to open a new PR?

**Comment by joecummings** [comment]
> Hello, this is great! Thank you, this helped me to get full DPO working, and I ended up just completing the remaining actions in the PR with training examples, configs, and tests. Would it make sense for me to open a new PR?

Yes please! Tag me on the PR for a review :)
